include/ssm.h

```diff
@@ -633,6 +380,215 @@ void ssm_desensitize(ssm_trigger_t *trig);
 
 /** @} */
 
+/**
+ * @addtogroup mem
+ * @{
+ */
+
+/** @brief Magic @a val_count value indicating a heap object is a builtin. */
+#define SSM_BUILTIN 0
+
+/** @brief Built-in types that are stored in the heap.
+ *
+ *  Type enumerated here are chosen because they cannot be easily or efficiently
+ *  expressed as a product of words. For instance, 64-bit timestamps cannot be
+ *  directly stored in the payload of a regular heap object, where even-numbered
+ *  timestamps may be misinterpreted as pointers.
+ */
+enum ssm_builtin {
+  SSM_TIME_T = 1, /**< 64-bit timestamps, #ssm_time_t */
+  SSM_SV_T,       /**< Scheduled variables, #ssm_sv_t */
+};
+
+/** @brief The metadata accompanying any heap-allocated object.
+ *
+ *  When @a val_count is 0, @a tag is to be interpreted as an #ssm_builtin.
+ */
+struct ssm_mm_header {
```

Comment:
Wart: I made this a separate struct to make it slightly easier to compute the size, but the padding issue makes me think twice about this choice.